drivers/spi/spi_stm32.c

```diff
@@ -459,129 +500,85 @@ static int spi_stm32_init(struct device *dev)
-#endif
-
-static const  struct spi_stm32_config spi_stm32_cfg_3 = {
-	.spi = (SPI_TypeDef *) CONFIG_SPI_3_BASE_ADDRESS,
-	.pclken = {
-		.enr = LL_APB1_GRP1_PERIPH_SPI3,
-		.bus = STM32_CLOCK_BUS_APB1
-	},
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-	.irq_config = spi_stm32_irq_config_func_3,
-#endif
-};
-
-static struct spi_stm32_data spi_stm32_dev_data_3 = {
-	SPI_CONTEXT_INIT_LOCK(spi_stm32_dev_data_3, ctx),
-	SPI_CONTEXT_INIT_SYNC(spi_stm32_dev_data_3, ctx),
-};
-
-DEVICE_AND_API_INIT(spi_stm32_3, CONFIG_SPI_3_NAME, &spi_stm32_init,
-		    &spi_stm32_dev_data_3, &spi_stm32_cfg_3,
-		    POST_KERNEL, CONFIG_SPI_INIT_PRIORITY,
-		    &api_funcs);
-
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-static void spi_stm32_irq_config_func_3(struct device *dev)
-{
-	IRQ_CONNECT(CONFIG_SPI_3_IRQ, CONFIG_SPI_3_IRQ_PRI,
-		    spi_stm32_isr, DEVICE_GET(spi_stm32_3), 0);
-	irq_enable(CONFIG_SPI_3_IRQ);
-}
-#endif
+/**
+ * @code{.codegen}
+ * _prio = codegen.config_property('CONFIG_SPI_INIT_PRIORITY', 70)
+ * _level = 'POST_KERNEL'
+ * _isr = 'spi_stm32_isr'
+ * _device_init = 'spi_stm32_init'
+ * _api = 'spi_stm32_driver_api'
+ * for x in range(0, int(codegen.device_tree_controller_count('SPI', 0))):
+ *     if not codegen.device_tree_controller_compatible_x('SPI', x, compatible):
+ *         continue
+ *     _base_address = codegen.device_tree_controller_property(
+ *                                 'SPI', x, 'BASE_ADDRESS')
+ *     _irq = codegen.device_tree_controller_property('SPI', x, 'IRQ_0')
+ *     _irq_prio = codegen.device_tree_controller_property(
+ *                                 'SPI', x, 'IRQ_0_PRIORITY')
+ *     _clock_enr = codegen.device_tree_controller_property(
+ *                                 'SPI', x, 'clocks_BITS')
+ *     _clock_bus = codegen.device_tree_controller_property(
+ *                                 'SPI', x, 'clocks_BUS')
+ *     _config_irq = codegen.device_tree_controller_data_name(
+ *                                 'SPI', x, 'config_irq')
+ *     _config_info = codegen.device_tree_controller_data_name(
+ *                                 'SPI', x, 'config')
+ *     _data = codegen.device_tree_controller_data_name('SPI', x, 'data')
+ *     _dev_name = codegen.device_tree_controller_device_name('SPI', x)
+ *     _drv_name = codegen.device_tree_controller_driver_name('SPI', x)
+ *     #
+ *     # config irq
+ *     if codegen.config_property('CONFIG_SPI_STM32_INTERRUPT'):
+ *         codegen.outl('DEVICE_DECLARE({});'.format(_dev_name))
+ *         codegen.outl('static void {}(void)'.format(_config_irq))
+ *         codegen.outl('{')
+ *         codegen.outl('\tIRQ_CONNECT( \\')
+ *         codegen.outl('\t\t{}, \\'.format(_irq))
+ *         codegen.outl('\t\t{}, \\'.format(_irq_prio))
+ *         codegen.outl('\t\t{}, \\'.format(_isr))
+ *         codegen.outl('\t\tDEVICE_GET({}), \\'.format(_dev_name))
+ *         codegen.outl('\t\t0);')
+ *         codegen.outl('\tirq_enable({});'.format(_irq))
+ *         codegen.outl('}')
```

Comment:
As this kind of block will be found in many other drivers, would there be a way to make it as an external codegen extension/function? 

We would just call codegen.device_declare(... with necessary parameters ...)

Reply:
Shure, it can be made a codegen function. Added that to my todo list.